Add ABFSS support to Ray Tune and Train

gangsf opened the pull request:
- Add _create_abfss_filesystem() helper function to handle Azure Blob File System Secure URIs
- Update get_fs_and_path() in both storage modules to detect and handle abfss:// URIs
- Integrate adlfs.AzureBlobFileSystem with PyArrow's PyFileSystem for compatibility
- Update checkpoint module to support ABFSS URIs in Checkpoint.from_uri()
- Update test files to handle ABFSS URIs when inspecting storage paths
- Fixes NotADirectoryError when using abfss:// storage paths with Ray Tune

Requires: pip install adlfs azure-identity

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Add ABFSS (abfss://) support by integrating adlfs with PyArrow, updating path resolution, and adapting checkpoints/tests to handle Azure Blob Storage (ADLS Gen2).
> 
> - **Storage**:
>   - Add `'_create_abfss_filesystem'` to parse/validate `abfss://` URIs and return a PyArrow `FileSystem` via `adlfs.AzureBlobFileSystem`.
>   - Update `get_fs_and_path` in `ray.train._internal.storage` and `ray.train.v2._internal.execution.storage` to detect `abfss://` and use the new helper.
> - **Checkpoint**:
>   - Update `ray.train._checkpoint.Checkpoint.__init__` to resolve `abfss://` using `'_create_abfss_filesystem'`.
> - **Tests**:
>   - Replace direct `pyarrow.fs.FileSystem.from_uri` calls with `get_fs_and_path` where appropriate.
>   - Adjust inspection helpers in tests to work with `abfss://` paths.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 44345859129df0ecc9002d75df5c96df78b9a1a2. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
Add ABFSS support to Ray Tune and Train

- Add _create_abfss_filesystem() helper function to handle Azure Blob File System Secure URIs
- Update get_fs_and_path() in both storage modules to detect and handle abfss:// URIs
- Integrate adlfs.AzureBlobFileSystem with PyArrow's PyFileSystem for compatibility
- Update checkpoint module to support ABFSS URIs in Checkpoint.from_uri()
- Update test files to handle ABFSS URIs when inspecting storage paths
- Fixes NotADirectoryError when using abfss:// storage paths with Ray Tune

Requires: pip install adlfs azure-identity
Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

justinvyu commented:
Hey @gangsf, thanks for the contribution. pyarrow added native support for abfss in `pyarrow==20.0.0`: https://arrow.apache.org/docs/python/generated/pyarrow.fs.AzureFileSystem.html#pyarrow.fs.AzureFileSystem

Can you try to see if that works out of the box with an upgraded pyarrow version?

Commit:
Consolidate and simplify ABFSS filesystem handling

- Consolidate _create_abfss_filesystem into ray.train._internal.storage
- Remove duplicate implementations from _checkpoint.py and v2 storage.py
- Simplify URI parsing using urlparse.username and urlparse.hostname
- Use get_fs_and_path helper in tests to reduce duplication
- Update error messages to cover both Train and Tune usage
- Apply formatting fixes (trailing whitespace)

Changes:
- Source files: -155 lines, +17 lines (138 lines removed)
- Test files: -18 lines, +6 lines (12 lines removed)
- Total: -173 lines, +23 lines (150 lines of duplicate code removed)

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

gangsf commented:
> Hey @gangsf, thanks for the contribution. pyarrow added native support for abfss in `pyarrow==20.0.0`: https://arrow.apache.org/docs/python/generated/pyarrow.fs.AzureFileSystem.html#pyarrow.fs.AzureFileSystem
> 
> Can you try to see if that works out of the box with an upgraded pyarrow version?

Tried both 20.0.0 and 21.0.0. Both got the error

```
File [~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:19](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:19)
      [1](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:1) # Licensed to the Apache Software Foundation (ASF) under one
      [2](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:2) # or more contributor license agreements.  See the NOTICE file
      [3](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:3) # distributed with this work for additional information
   (...)
     [15](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:15) # specific language governing permissions and limitations
     [16](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:16) # under the License.
---> [19](https://vscode-remote+vscode-002dsession-002dh6kqg54wm8lv6lz62d62iseemd-002ei-002eanyscaleuserdata-002dstaging-002ecom.vscode-resource.vscode-cdn.net/home/ray/default/~/anaconda3/lib/python3.12/site-packages/pyarrow/json.py:19) from pyarrow._json import ReadOptions, ParseOptions, read_json, open_json  # noqa

ImportError: cannot import name 'open_json' from 'pyarrow._json' (/home/ray/anaconda3/lib/python3.12/site-packages/pyarrow/_json.cpython-312-x86_64-linux-gnu.so)

During handling of the above exception, another exception occurred:
```

matthewdeng commented:
@gangsf can you try to resolve the error? That seems unrelated to abfss.
